src/Logger/Adapter/Raygun.php

```diff
@@ -0,0 +1,161 @@
+            \array_push($breadcrumbsArray, [
+                'category' => $breadcrumb->getCategory(),
+                'message' => $breadcrumb->getMessage(),
+                'type' => $breadcrumb->getType(),
+                'level' => "request",
+                'timestamp' =>\intval($breadcrumb->getTimestamp())
+            ]);
+        }
+
+        $tagsArray = [];
+
+        foreach($log->getTags() as $tagKey => $tagValue) {
+            \array_push($tagsArray, $tagKey . ': ' . $tagValue);
+        }
+
+        \array_push($tagsArray, 'type: ' . $log->getType());
+        \array_push($tagsArray, 'environment: ' . $log->getEnvironment());
+        \array_push($tagsArray, 'sdk: utopia-logger/' . Logger::LIBRARY_VERSION);
+
+        // prepare log (request body)
+        $requestBody = [
+            'occurredOn' =>  \intval($log->getTimestamp()),
+            'details' => [
+                'machineName' => $log->getServer(),
+                'groupingKey' => $log->getNamespace(),
+                'version' => $log->getVersion(),
+                'error' => [
+                    'className' => $log->getAction(),
+                    'message' => $log->getMessage()
+                ],
+                'tags' => $tagsArray,
+                'userCustomData' => $log->getExtra(),
+                'user' => [
+                    'isAnonymous' => empty($log->getUser()),
+                    'identifier' => empty($log->getUser()) ? null : $log->getUser()->getId(),
+                    'email' => empty($log->getUser()) ? null : $log->getUser()->getEmail(),
+                    'fullName' => empty($log->getUser()) ? null : $log->getUser()->getUsername(),
+                ],
+                'breadcrumbs' => $breadcrumbsArray
+            ]
+        ];
+
+        // init curl object
+        $ch = curl_init();
+
+        // define options
+        $optArray = array(
+            CURLOPT_URL => 'https://api.raygun.com/entries',
+            CURLOPT_RETURNTRANSFER => true,
+            CURLOPT_POST => true,
+            CURLOPT_POSTFIELDS => \json_encode($requestBody),
+            CURLOPT_HEADEROPT => CURLHEADER_UNIFIED,
+            CURLOPT_HTTPHEADER => array('Content-Type: application/json', 'X-ApiKey: ' . $this->apiKey)
+        );
+
+        // apply those options
+        curl_setopt_array($ch, $optArray);
+
+        // execute request and get response
+        $result = curl_exec($ch);
+
+        // also get the error and response code
+        $errors = curl_error($ch);
+        $response = curl_getinfo($ch, CURLINFO_HTTP_CODE);
+
+        curl_close($ch);
+
+        return $response;
+    }
+
+    /**
+     * Raygun constructor.
+     *
+     * @param string $configKey
+     */
+    public function __construct(string $configKey)
+    {
+        $this->apiKey = $configKey;
+    }
+
+    public function validateLog(Log $log): bool
+    {
+        // Support all log types (as tag)
+        switch ($log->getType()) {
```

Comment:
If this is a repeating trend, I think we should have an array containing the allowed log types per adapter. Then we can implement `isLogSupported` method (or a better name) on the abstract adapter directly. This method could also be useful for dev using the lib, it can allow them to know what might be supported before getting errors.

Reply:
Not sure if everything can be implemented like this. For example, we would need `getSupportedLogTypes()`, `getSupportedBreadcrumbTypes()` and `getSupportedEnvironmentTypes()`.

- Some providers might support `any` environment type, so the function would have to return some hack-y value to say "anything supported".
- Some providers might not support breadcrumbs, so it's silly to try to return an array of supported breadcrumb types. 
- Some providers might have more options than we support, for example required field for the device name. How does a developer achieve that? Well, he would need to use `tags` or `extra` array to provide that information. Then, adapter needs to nave this special check where he looks at `tags` array and check if `deviceName` is present.

Reply:
@eldadfux currently I don't think it would be possible to make this as part of the abstract class since not all providers may have concepts like `breadcrumbType`, `logType`, `environment` . So we decided to use a generic `validateLog()` method so each provider can have a different way to validate their log. 

As we have more providers, we can better decide which methods need to make way to the abstract class. Let us know what you think

Reply:
Implemented `getSupportedTypes`, `getSupportedEnvironments`, `getSupportedBreadcrumbTypes` in https://github.com/utopia-php/logger/commit/74140bf8bf9cb8aeb19f7505506e9cd9b71b725b